geode-core/src/test/java/org/apache/geode/cache/query/internal/AttributeDescriptorTest.java

```diff
@@ -232,18 +307,91 @@ public void readReflectionForImplicitMethodInvocationShouldThrowExceptionWhenMet
     AttributeDescriptor attributeDescriptor =
         new AttributeDescriptor(typeRegistry, methodInvocationAuthorizer, attributeName);
 
-    assertThatThrownBy(() -> attributeDescriptor.readReflection(testBean))
+    assertThatThrownBy(() -> attributeDescriptor.readReflection(testBean, queryExecutionContext))
+        .isInstanceOf(NotAuthorizedException.class)
+        .hasMessageStartingWith(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING);
+  }
+
+  @Test
+  @Parameters({
+      "nonPublicAttributeWithPublicAccessor, nonPublicAttributeWithPublicAccessor",
+      "nonPublicAttributeWithPublicGetterMethod, getNonPublicAttributeWithPublicGetterMethod"})
+  public void readReflectionForImplicitMethodInvocationShouldShouldThrowExceptionWhenMethodIsNotAuthorizedAndCacheResult(
+      String attributeName, String methodName) {
+    doReturn(false).when(methodInvocationAuthorizer).authorize(any(), any());
+    AttributeDescriptor attributeDescriptor =
+        new AttributeDescriptor(typeRegistry, methodInvocationAuthorizer, attributeName);
+
+    assertThatThrownBy(() -> attributeDescriptor.readReflection(testBean, queryExecutionContext))
         .isInstanceOf(NotAuthorizedException.class)
         .hasMessageStartingWith(RestrictedMethodAuthorizer.UNAUTHORIZED_STRING);
+
+    String cacheKey = TestBean.class.getCanonicalName() + "." + methodName;
+    assertThat(queryExecutionContext.cacheGet(cacheKey)).isNotNull();
+    assertThat(queryExecutionContext.cacheGet(cacheKey)).isEqualTo(false);
+  }
+
+  @Test
+  public void readReflectionForImplicitMethodInvocationShouldUseAlreadyAuthorizedCachedResultWhenMethodIsForbiddenAndQueryContextIsTheSame() {
```

Comment:
The "already authorized" part of this test name is a little confusing. Since the authorizer is returning false for all methods, none of them have been "authorized" in the sense of being allowed, just in the sense of having been passed to the `authorize()` method. I think readReflectionForImplicitMethodInvocationShouldUse**CachedAuthorizerResult**WhenMethodIsForbiddenAndQueryContextIsTheSame might be a clearer name.

Reply:
👍